Add post about ggplot2 migrating to S7 (#735)
tidyverse/tidyverse.org: content/blog/ggplot2-4-0-0-s7/index.Rmd

The way to add additional information to classes in S7 is to make a subclass with additional properties.
For example, if we want to add colour information to a new plot, we can do the following:

```{r}
inked_ggplot <- S7::new_class(
  name = "inked_ggplot",
  parent = class_ggplot, 
  properties = list(ink = S7::class_character)
)

inked_ggplot
```

When you define a new class, the object you've assigned it to automatically becomes the class definition which comes with a free, standard constructor.
This means that we can start building new plots with our subclass right away.
Note that we haven't implemented any behaviour around the `ink` property (yet), so it will just print like a normal plot.

```{r}
my_plot <- inked_ggplot(data = mpg, ink = "red") +
  geom_point(aes(displ, hwy))
my_plot
```

In contrast to S3, where you would change list-items by using `$`, in S7 you can use `@` to read and write properties.
So if we want to change the stored `ink` colour, we can use:

```{r}
my_plot@ink <- "blue"
```

### Testing

In S3, the recommended way to test for the class of an object is to use a testing function.

> For example `is.factor()`; and if such a testing function doesn't exist: use `inherits()`.

Reviewer: ```suggestion
An example is `is.factor()` but it may be that such a testing function doesn't exist:. In that case you can use `inherits()`.
```

```{r}
testthat::test_that(
  "the plot object has the ggplot class",
  {
    plot <- ggplot()
    
    # Works regardless of S3 or S7
    testthat::expect_true(is_ggplot(plot))
    
    # This will become dysfunctional in the future.
    # Do not use this!
    testthat::expect_s3_class(plot, "ggplot")
    
    # This will work in the new version
    testthat::expect_s7_class(plot, class_ggplot)
  }
)
```

The ggplot2 package manually appends the `"ggplot"` class for backwards compatibility reasons (likewise for `"theme"`).
However, once this phases out, the `testthat::expect_s3_class()` expectation will become untenable.
It is also currently flawed, as it does not work for subclasses!

```{r, error=TRUE}
testthat::test_that(
  "the inked plot has the ggplot class",
  {
    plot <- inked_ggplot()
    testthat::expect_s3_class(plot, "ggplot")
  }
)
```

The advice herein is thus to use `is_ggplot()`.

## Generics and methods

If you are new to object oriented programming in R, you might be unfamiliar what the terms 'generic' and 'methods' mean.

> They are a form of 'polymorphism', where we can use a single function, called the 'generic' function, with different implementations for different classes (where one such implementation is called a 'method').

Reviewer: ```suggestion
They are a form of 'polymorphism', that allow us to use a single function, called the 'generic' function, with different implementations for different classes (where one such implementation is called a 'method').
```

A well known generic is `print()`, which does different things for different classes. 

> For example `print(1:10)` prints the numeric vector to the console, but `print(my_plot)` opens a graphics device to render the plot.

Reviewer: ```suggestion
For example `print(1:10)` prints the numeric vector to the console, but `print(my_plot)` opens a graphics device and renders the plot.
```
